File: python_part/Leetcode/DP/Easy/198. House Robber(DP)/DP.py
class Solution:
    def rob(self, nums) -> int:
        size = len(nums)
        if size == 0:
            return 0
        if size == 1:
            return nums[0]
        if size == 2:
            return max(nums[0], nums[1])

        dp = [None for _ in range(size)]
        dp[0] = nums[0]
        dp[1] = max(nums[1],nums[0])

        for i in range(2,size):
            dp[i] = max(dp[i-2]+nums[i], dp[i-1])

        return dp[-1]

foo = Solution()
foo.rob([2,1,1,2])

File: Cpp_part/Leetcode/Leetcode/DP/面试题42. 连续子数组的最大和（最大公共子序）/dp.cpp
#include <iostream>
#include <vector>

using namespace std;
/*
dp[i]表示以i结尾的最大的和，必须要包含i
 */
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        vector<int> dp = {0};
        int maxs =  nums[0];
        for(int i = 0; i < nums.size(); i ++){
            dp.push_back(max(dp[i]+nums[i], nums[i]));
            maxs = max(dp[i+1], maxs);
        }
        return maxs;
    }
};
